### PR TITLE
NFT as an account (NFTAA)

### PR DESCRIPTION
**Pull request by Roman-24**
#### Project Abstract

NFT as an Account (NFTAA) is an innovative project aimed at revolutionizing the use of NFTs within the blockchain ecosystem, particularly in the Substrate based networks like Polkadot and Kusama. The project seeks to leverage the unique properties of NFTs to function as proxy accounts, enhancing interoperability, security, and flexibility across various blockchain networks. By enabling NFTs to hold assets, stake, and interact with other blockchain components similarly to standard accounts, NFTAA introduces a novel approach to blockchain interactions.

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [x] **Level 2**:  Up to $30,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] Payment details have been provided (Polkadot AssetHub (DOT, USDC & USDT) address in the application and bank details via email, if applicable).
- [x] I understand that an agreed upon percentage of each milestone will be paid in vested DOT, to the Polkadot address listed in the application.
- [x] I am aware that, in order to receive a grant, I (and the entity I represent) have to successfully complete a KYC/KYB check.
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

### CODE REVIEW

**Comment by Roman-24** [comment]
I have read and hereby sign the Contributor License Agreement.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the application. Just one quick feedback here:  The default deliverables 0a-0d of the [template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) are mandatory for all milestones and deliverable 0e at least for the last one.

**Comment by Roman-24** [comment]
@Noc2 thank you for your feedback. Our application has been modified and default deliverables 0a-0d have been added to milestones 2 and 3. The point about the article in milestone 3 has been modified from 3a to 0e to make it consistent with the templates. We believe that our application now complies with all the necessary formalities.

**Review by semuelle — COMMENTED** [review]
Thank you for the application and the long wait, @Roman-24. 

This looks like an interesting idea. However, I'm a bit wary of another NFT pallet. We have supported several NFT implementations, each with their own twist, but few have actually been integrated into parachains or seen much adoption. Can you elaborate what your intention is here? From the looks of it, this is more of an "applied research" exercise than anything.

**Comment by Roman-24** [comment]
@semuelle Thank you for your feedback.

Our goal is to expand the boundaries of how web3 uses NFTs today. It's an extension of how current NFTs are used, adding a whole new set of capabilities. Another feature addresses the problem of current account illiquidity. If you want to change the owner of an account (or, for example, a locked staking), there is no secure way to do it (except by sharing private keys (unsecure thing)). In our proposal, we are primarily focusing on the use case within staking, but there are several other use cases, which we have addressed in the "Business Overview" subsection. The implementation of this idea makes sense and is supported by the fact that there is a similar proposal for Ethereum, which we have discussed in the "Differences and Distinctions (NFTAA vs. EIP-6551)" paragraph.

As mentioned in the proposal, we have already conducted research on this topic, along with some prototypes. Therefore, we now want to move towards a real implementation that will provide value to real users.
We would also like to point to the ongoing cross-chain NFT development, which in the future, combined with NFTAA, could enable sendable accounts across the entire Polkadot Paraverse.

**Review by semuelle — APPROVED** [review]
> we now want to move towards a real implementation that will provide value to real users.

Thanks for the clarification, @Roman-24. My question was more geared towards understanding whether you are planning to deploy this yourself or if you have parachain teams interested in it. But either way, it looks like an interesting project to me, so I'm giving my approval and will share it with the rest of the committee.

In the meantime, could you fill out the [KYB form](https://in.sumsub.com/idensic/l/#/uni_JRqwo72Y5AjXaRDO) (assuming you are applying as a company, otherwise please use [this form](https://in.sumsub.com/idensic/l/#/uni_mjcB05gJZfh8dbr7))?

**Comment by Roman-24** [comment]
We have completed the KYB and therefore verification is in process. We are glad you like our idea and thank you for your support @semuelle. We will definitely communicate with the parachains, we believe we have a good network for that.

File: applications/nftaa.md [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+unique concepts introduced by NFTAA.
+
+The NFTAA project addresses several critical needs within the blockchain ecosystem.
+Utilizing NFTAA provides more methods for managing and delegating staking rights without exposing private keys.
+This approach significantly reduces the risk of unauthorized access and enhances overall security.
+Furthermore, NFTAA allows staked assets to remain liquid, enabling users to trade, lend, or use them as collateral even
+while they are staked.
+This flexibility is crucial for users who need access to their assets without compromising their staking benefits.
+The project also potentially leverages Polkadot's ability to connect multiple blockchains, ensuring seamless asset transfers and
+interactions across different networks.
```
Comment (PieWol) [inline applications/nftaa.md:140]: How exactly will this solution work in a cross chain style? As you are not the holder of the private key to the address of the keyless account you couldn't access the funds on the account with the same address on another network. This sounds like a problem to me if you were to simply transfer funds via xcm. How are you solving this?

File: applications/nftaa.md [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+  amount.
+- Bifrost: Provides liquidity for staked assets through a voucher system and supports cross-chain staking.
+- Acala: Offers a similar staking mechanism with its LDOT token, providing immediate liquidity and flexibility in
+  staking.
+
+Differences and Distinctions:
+
+- NFTAA vs. Lido/Bifrost/Acala: Unlike these projects, NFTAA focuses on using NFTs as proxy accounts for staking, which
+  allows for a more flexible and secure delegation of staking rights and ownership. NFTAA also leverages the composability
+  of NFTs, enabling them to hold other assets and act as accounts within the blockchain ecosystem.
```
Comment (PieWol) [inline applications/nftaa.md:166]: How is this more secure? Because you don't rely on the business logic of a party like Acala but instead you rely on the business logic of yours in the form of this very pallet?

File: applications/nftaa.md [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+Additionally, NFTAA’s focus on using NFTs as proxy accounts for staking and asset management offers a novel approach compared to the more traditional liquid staking and smart contract enhancements in the Ethereum ecosystem.
+
+EIP 6551 presents certain limitations. Our solution, e.g., does not prioritize connecting existing NFTs but enables bidirectional binding of the NFT and proxy.
+Consequently, upon ownership of an NFT, we can discern from the NFT metadata indicating its status as an NFTAA.
+Furthermore, we enhance transparency by ensuring the atomic creation of the NFT and proxy within a single transaction, thereby promoting seamless integration.
+Likewise, our design does not need the registrar service to create the TBA address as opposed to TBA.
+In our case, the pallet creates the NFTAA with functionality that also works as the NFTAA address (TBA uses one more address than NFTAA).
+Another difference is that in TBA, you could have more accounts on one NFT, which, in our case, is not possible due to the atomic creation of exactly one NFT at the time of creating the NFTAA.
```
Comment (PieWol) [inline applications/nftaa.md:184]: The whole substrate runtime is synchronous so I don't see why this "atomic creation" would be unique to your solution.

File: applications/nftaa.md [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+  allows for a more flexible and secure delegation of staking rights and ownership. NFTAA also leverages the composability
+  of NFTs, enabling them to hold other assets and act as accounts within the blockchain ecosystem.
+
+Similar concepts exist in other ecosystems:
+
+- Rocket Pool: An Ethereum-based protocol for decentralized staking that allows users to earn rewards without running a
+  full node.
+- EIP-6551 (TBA - token bound accounts): A proposed Ethereum standard to enhance NFTs by equipping them with smart contract capabilities, allowing them to own assets and interact with other contracts.
+
+Differences and Distinctions (NFTAA vs. EIP-6551):
```
Comment (PieWol) [inline applications/nftaa.md:174]: Why is an EIP a key reference here, this is Substrate after all and not an EVM chain which we are talking about. Without defining how you translate the EIP to a Substrate equivalent this is hard to follow.

File: applications/nftaa.md [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+In this case, if the upcoming smart contract is badly developed (does not contain an execute function for byte calls to other contracts) or is not upgradeable, it can happen that assets or liquidity sent to the TBA will not be withdrawable and become locked forever (even the NFT bounded in TBA can be sent by a mistake (if not safeguarded) to the address of TBA. Therefore you lock yourself from it).
+The problem is that TBA uses only the address of any contract you want in the registry contract and does not look for any details on how the contract looks or if it is already deployed. In our case, the account form is pre-defined in the factory pallet, and even after deployment, we guarantee upgradeability as it is a proxy account.
+The second security issue is potential fraud, which is [described in the EIP 6551 specification](https://eips.ethereum.org/EIPS/eip-6551\#fraud-prevention) that you withdraw assets from the TBA and at the same time sell the NFT to someone who thought that they would also have the assets in the TBA. In our case, we provide the security measures for not withdrawing the assets from the NFTAA while also selling the NFT, which is bound to the NFTAA.
```
Comment (PieWol) [inline applications/nftaa.md:188]: Will these security measures be included in your currently developed pallet? I don't see it in the milestones.

File: applications/nftaa.md [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+- pallet::dispatchables::set_attribute
+- pallet::dispatchables::set_collection_max_supply
+- pallet::dispatchables::set_collection_metadata
+- pallet::dispatchables::set_metadata
+- pallet::dispatchables::set_price
+- pallet::dispatchables::set_team
+- pallet::dispatchables::transfer
+- pallet::dispatchables::transfer_ownership
+- pallet::dispatchables::unlock_item_transfer
+- pallet::dispatchables::update_mint_settings
```
Comment (PieWol) [inline applications/nftaa.md:313]: Would you mind to describe these functions? Deducting meaning from function names can easily lead to misunderstandings. Thanks.

File: applications/nftaa.md [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+| **0c.** | Testing and Testing Guide                    | Core functions will be fully covered by comprehensive unit tests to ensure functionality and robustness. In the guide, we will describe how to run these tests. |
+| **0d.** | Docker                                       | We will provide a Dockerfile(s) that can be used to test all the functionality delivered with this milestone. |
+|    2a. | nftaa_check             | We will add functionalities to read and check if NFTAA exists for a given account                                                  |
+|    2b. | nftaa_collections       | We will add functionalities to read existing collections and manage NFTAAs if they are in a collection.                            |
+|    2c. | nftaa_stake/unstake     | The functionality needed to do stake/unstake operations on behalf of NFTAA, its like deposit/withdrawal functions                  |
+|    2d. | nftaa_increase/decrease | Functions needed to increase or decrease stake by NFTAA, this is needed for full compatibility with traditional methods of staking |
```
Comment (PieWol) [inline applications/nftaa.md:366]: I'm confused by this milestone. Previously you described your project as a solution to use the NFTAAs as virtual keyless accounts which are able to do anything a regular account can. Yet this milestone sounds like you are putting in a lot of work to make your solution compatible with the existing pallets for staking. If this is indeed necessary I think you should try to find a more generic solution which doesn't create this overhead for every pallet you want to support with NFTAAs.

File: applications/nftaa.md [hunk 8]
```diff
@@ -0,0 +1,411 @@
...
+| **0d.** | Docker                                       | We will provide a Dockerfile(s) that can be used to test all the functionality delivered with this milestone. |
+|    0e. | Article         | We will publish an **article** that explains NFTAA (what was done/achieved as part of the grant). The article will be published through an online media service, e.g., medium.com, and advertised through social networks. Our ambition, as we are from academic environment, is to have also an academic paper about the idea of NFTAA in Polkadot published in a conference venue                                                                                    |
+|    3a. | Web app         | Develop a web application to interact with the nftaa_pallet, allowing users to create, manage, and transfer NFTAAs, provide a web application to interact with the nftaa_pallet: allowing users to create collections and single item; manage of metadata and attributes; operate with NFTAA ownership (transfers); play with staking, increase stake and unstake handled by NFTAA; read operations as get for an account, get assets of NFTAA, get the owner of NFTAA |
```
Comment (PieWol) [inline applications/nftaa.md:381]: We already have a ton of good staking interfaces and webapps. I don't think it makes sense to reinvent the wheel here. Instead I would suggest to integrate your necessary changes to a popular webapp like the Polkadot Staking Dashboard to make staking for NFTAAs work in a convenient way.

File: applications/nftaa.md [hunk 9]
```diff
@@ -0,0 +1,411 @@
...
+|    3b. | Marketplace app | We will also create a lightweight marketplace application for easily buying and selling the NFTAAs with an overall list in a form of a bulleting board.                                                                                                                                                                                                                                                                                                                |
```
Comment (PieWol) [inline applications/nftaa.md:382]: Why is this necessary and wouldn't be supported by existing NFT marketplaces?

**Review by PieWol — CHANGES_REQUESTED** [review]
Thanks for your interest in the Grants Program! :)
I left some comments and am looking forward to your statements and answers.

File: applications/nftaa.md [hunk 10]
```diff
@@ -0,0 +1,411 @@
...
+
+Our motivation for creating this project stems from the desire to address the inherent limitations and inefficiencies
+associated with traditional staking methods.
+Issues such as the illiquidity of staked assets and the complexities involved in transferring ownership pose significant
+challenges.
+By utilizing NFTs as accounts, we aim to provide a more secure and flexible staking solution, enhancing user control and
+ownership.
+This innovative approach opens new avenues for business opportunities and applications in DeFi
+and enterprise blockchain solutions, underscoring our commitment to advancing blockchain technology and its
+applications.
```
[the remark this answers is not in the document]
Comment (Roman-24) [reply applications/nftaa.md:34]: It offers a different approach and provides an alternative option. At no point do we claim that it is superior or inferior to conventional liquid staking solutions. While our NFT proxy accounts enable us to address more than just the illiquidity of staked assets (e.g. digital assets management, company relations management), unlike traditional liquid staking derivatives, we present it as a complementary solution.

File: applications/nftaa.md [hunk 11]
```diff
@@ -0,0 +1,411 @@
...
+Through forkless runtime upgrades, it is possible to enhance the functionality of individual modules.
+This flexibility is critical as runtime pallets operate close to the system’s core, offering virtually limitless
+modification options.
+Although the transaction weight is a limiting factor in terms of computing resources available for each transaction, it
+does not detract from the Substrate framework’s overall power and adaptability.
+
+Most Substrate pallets are designed based on a single account architecture, aligning with the NFTAA philosophy and
+facilitating smooth integration with existing infrastructure.
+By treating NFTAAs as conventional accounts, we can easily leverage the existing infrastructure for seamless
+communication.
```
[the remark this answers is not in the document]
Comment (Roman-24) [reply applications/nftaa.md:59]: We're glad you see it the same way.

File: applications/nftaa.md [hunk 12]
```diff
@@ -0,0 +1,411 @@
...
+Additionally, NFTAAs can function as asset owners or participate in multi-signature schemes, further
+enhancing their utility and security.
```
[the remark this answers is not in the document]
Comment (Roman-24) [reply applications/nftaa.md:61]: We see this as an opportunity for the security of multi-sig schemes so that NFTAA can change the owner. If we have an ordinary account with a leaked PK there is nothing we can do about it. If we have an NFTAA and the owner of such an account leaks PK it can change the owner (i.e. PK) tied to the NFTAA and thus protect the multi-sig (assuming it makes it before the accountant does). But we acknowledge that this may be controversial and we would probably just write utility instead of utility and security :)

File: applications/nftaa.md [hunk 13]
```diff
@@ -0,0 +1,411 @@
...
+meaning, that Bob will have full access using his own private keys that only he knows.
+This gives NFTAA the unique ability to assess many complex use cases.
+
+#### Technical Details
+
+The implementation of the NFTAA concept within the Polkadot ecosystem can be approached in two primary ways.
+The first approach involves creating custom blockchain logic in the form of a pallet, written in Rust as a Substrate
+module.
+The second approach involves using smart contracts. For this proposal, we prefer the pallet approach due to its
+integration capabilities and performance benefits.
```
[the remark this answers is not in the document]
Comment (Roman-24) [reply applications/nftaa.md:78]: We are glad that you agree with our decision of tech details.
According to our information there are some possibilities to do smart contracts, but we think it's only possible on some parachains (https://wiki.polkadot.network/docs/build-smart-contracts). 
Anyway, if there will be NFTAA pallet in the future it will be possible to add chain extensions and bring this functionality to smart contract layer.

File: applications/nftaa.md [hunk 14]
```diff
@@ -0,0 +1,411 @@
...
+
+The proposed state-of-the-art pallet is based on the _nfts_ and _utility_ pallets.
+The first step involves implementing and wrapping the NFT logic from the _nfts_ pallet into the _nftaa_ pallet.
+A collection in the _nfts_ runtime pallet is created to represent an equivalent of a smart contract instance.
+The _nftaa_ pallet marks a collection so that any token from this collection is treated as an account. After creating a
+mintable collection, the _nftaa_ special function must be called with the appropriate parameters, including _collection_id_
+and _attributes_.
+Attributes are crucial for enterprise use cases, allowing NFTAAs to be marked as entities such as companies or departments
+within a corporate structure.
```
[the remark this answers is not in the document]
Comment (Roman-24) [reply applications/nftaa.md:87]: From our point of view, this is an opportunity to differentiate between already existing unique NFTs inside one company. To better manage and govern the orgs represented by NFTAAs or to add additional information to the NFT that is needed to implement the usecase we are writing about.

File: applications/nftaa.md [hunk 15]
```diff
@@ -0,0 +1,411 @@
...
+
+<img src="https://raw.githubusercontent.com/decenzio/NFTAA-Polkadot/main/proposal/nftaa_seq_dia_creation.svg" alt="Proposal Diagram" width="70%" height="30%">
+
+Reimplementing the _utility.as_derivate_ function is the second crucial step.
+Constructing a keyless address in the runtime pallet from accountId and index is challenging.
+The index must represent a concatenation of the collection and token ID, but since the index is a u16 type, it cannot
+accommodate two u32 types.
+To overcome this, we will use the _blake2_ hash function to encode the following properties: a binary prefix _"modlpy/nftproxy"_, followed by parachain, collection, and token ID. Both collection and token ID are u32 types, and
+padding them with zeros to length 32 prevents overlap.
+The _AccountId32_ type is then used to construct the address, ensuring a unique address for each NFTAA.
```
[the remark this answers is not in the document]
Comment (Roman-24) [reply applications/nftaa.md:97]: Maybe we dont understand what are you trying to address. However, a user can not manipulate counter of NFTs in collection, so generated address will be always unique.

File: applications/nftaa.md [hunk 16]
```diff
@@ -0,0 +1,411 @@
...
+also opens new possibilities for their application in various business and financial contexts.
+By combining the strengths of NFTs and blockchain technology, NFTAA represents a significant step forward in the
+evolution of digital assets and their management.
+
+### Ecosystem Fit
+
+The NFTAA project aims to modernize staking and account management within the Polkadot and Substrate ecosystems through
+the innovative use of NFTs by leveraging the flexibility of the Substrate framework and Polkadot's interoperability features.
+These NFT-based accounts can own assets, participate in staking, and serve as proxies for other accounts, significantly
+enhancing security, transparency, and liquidity within blockchain networks.
```
[the remark this answers is not in the document]
Comment (Roman-24) [reply applications/nftaa.md:119]: In principle, the example we gave above applies. However, in addition, we believe that having more options on how to approach asset management increases security as we can choose the way that suits us best.

In the same way, we believe that flexibility increases security.

We are open to discuss this aspect more :)

File: applications/nftaa.md [hunk 17]
```diff
@@ -0,0 +1,411 @@
...
+dApp developers will find NFTAA particularly useful for secure and flexible account management and staking solutions.
+Developers of cryptocurrency wallets can integrate NFTAA to provide users with advanced functionalities, thereby
+improving the user experience.
+Both individuals and enterprises looking for improved liquidity in their blockchain assets can leverage
+NFTAA for better asset management.
+Additionally, researchers and academics interested in exploring new blockchain innovations can study and expand upon the
+unique concepts introduced by NFTAA.
+
+The NFTAA project addresses several critical needs within the blockchain ecosystem.
+Utilizing NFTAA provides more methods for managing and delegating staking rights without exposing private keys.
```
[the remark this answers is not in the document]
Comment (Roman-24) [reply applications/nftaa.md:134]: If we have 1000 DOT locked in staking and suddenly we need to sell them for some reason we have to wait 28 days. If we want to sell the account the only option is to sell my PK (unsecure).

If we have 1000 DOTs staked through NFTAA we can simply sell the NFTAA (probably for any compensation) and it will be safe for both parties.

Staking is just an example, there are a number of different examples regarding locks and assets.

File: applications/nftaa.md (hunk 1 above)
Comment (Roman-24) [reply applications/nftaa.md:140]: We are looking into and elaborating more about at cross-chain NFTs idea what is being done by other teams (e.g. Paraspell or Unique Network) but currently you are 100% right. But we will see what will future bring to us.

File: applications/nftaa.md (hunk 2 above)
Comment (Roman-24) [reply applications/nftaa.md:166]: Safer from the user's point of view that this way we can bypass various time locks. So we don't expose our self to the danger of illiquidity. Or, as we mentioned before, we take more flexible as more secure.

We don't question Acala's shared security model as it is inherited from the relay chain, but in our opinion, the pallet is more testable than the whole parachain or at least quicker and easier testable. Therefore we think that the easiness to test the pallet makes it at the end more robust (and maybe also secure).

File: applications/nftaa.md (hunk 4 above)
Comment (Roman-24) [reply applications/nftaa.md:174]: This is just related work and mainly a comparison of business characteristics not a comparison of technicalities. We will not translate EIP to substrate equivalent.

File: applications/nftaa.md (hunk 3 above)
Comment (Roman-24) [reply applications/nftaa.md:184]: We didn't realize that, to be honest, and so you're right. But atomic creation guarantees same conditions across all parachains and future usages of pallet and moreover it is a guarantee to us that process works exactly as it was designed, so we take it as another safety catch.

File: applications/nftaa.md (hunk 5 above)
Comment (Roman-24) [reply applications/nftaa.md:188]: We will try to address this problem within existing features, but at this point we cannot exclude the possibility that this mechanism may need to be improved later.

File: applications/nftaa.md (hunk 6 above)
Comment (Roman-24) [reply applications/nftaa.md:313]: Yes, in order not to re-do already made and to ensure compatibility we want to use the NFT pallet as a dependency. We are ready to maintain compatibility in future.

File: applications/nftaa.md (hunk 7 above)
Comment (Roman-24) [reply applications/nftaa.md:366]: We don't want to do this for every pallet and we certainly want our solution to be generic, but the case of staking is special and therefore we want to have these functionalities explicitly ready. But it still applies NFTAAs are virtual keyless accounts which are able to do anything a regular account can.

File: applications/nftaa.md (hunk 8 above)
Comment (Roman-24) [reply applications/nftaa.md:381]: We like this idea and we agree with you. This milestone can be understood as developing the integration to existing solutions, we definitely do not want to reinvent the wheel.

File: applications/nftaa.md (hunk 9 above)
Comment (Roman-24) [reply applications/nftaa.md:382]: We can certainly make sales through the Koda platform, for example, and so it's true that we can prepare an integration to them instead of our own web app. We also have a good relationship with them and we believe that this kind of cooperation will not be a problem.

The lightweight application is more for testing purposes.

**Comment by Roman-24** [comment]
@PieWol thank you so much for your effort to review our proposal. 

We've tried to answer your questions as best we can. If there is still something that is not clear we will be happy to continue the discussion.

File: applications/nftaa.md (hunk 7 above)
Comment (PieWol) [reply applications/nftaa.md:366]: What makes staking special?

File: applications/nftaa.md (hunk 8 above)
Comment (PieWol) [reply applications/nftaa.md:381]: great, would be nice if you could adapt the wording to that strategy.

File: applications/nftaa.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (PieWol) [reply applications/nftaa.md:97]: I would like to see a mathematical proof that the user is not able to forge addresses that are associated with NFTAAs.

File: applications/nftaa.md (hunk 8 above)
Comment (Roman-24) [reply applications/nftaa.md:381]: We will definitely do it :)

File: applications/nftaa.md (hunk 7 above)
Comment (Roman-24) [reply applications/nftaa.md:366]: We believe staking is an ideal use case to showcase how NFTAA operates because it's a concept that many users are already familiar with, making it easier to grasp the potential of NFTAA. Additionally, staking plays a crucial role in maintaining network security, making it a fitting example to demonstrate the robustness and functionality of NFTAA in a practical environment.

File: applications/nftaa.md (hunk 8 above)
Comment (semuelle) [reply applications/nftaa.md:381]: Are you still working on this, @Roman-24?

Comment (Roman-24) [reply applications/nftaa.md]: Changes have been made in  [this commit](https://github.com/w3f/Grants-Program/pull/2396/commits/1a8e20e1d464cf4d7e29e3da464d0673f3ca191d)

**Comment by Roman-24** [comment]
@PieWol We have modified the milestone changes to say that we will integrate to existing applications. You can see the changes [here](https://github.com/w3f/Grants-Program/pull/2396/commits/1a8e20e1d464cf4d7e29e3da464d0673f3ca191d)

**Comment by PieWol** [comment]
Hey @Roman-24 ,
sorry for the delay here. I'll get back to your comments today and I'm sure the rest of the team will revisit your application soon. There is nothing missing on your side. Thanks.

File: applications/nftaa.md (hunk 9 above)
Comment (PieWol) [reply applications/nftaa.md]: Why do you need this for testing purposes? For your internal testing you can just use bare extrinsics. Once you made sure your NFTs are working as expected it should work like a charm with Koda. Why wouldn't your NFTs be tradeable just like any other NFTs?

File: applications/nftaa.md (hunk 7 above)
Comment (PieWol) [reply applications/nftaa.md:366]: I know what staking ist. I just don't understand your answer. If you are planning to make your solution generic as in "this NFTAA can be used just like a regular account", then why do you need to put in any work to make your solution compatible with the staking pallet. I hope you understand where my issue is here. Staking is already supported for regular accounts. With you adding just another layer on top of accounts with this form of account abstraction I don't understand why you would need to modify anything below this account level. E.g. How individual pallets work with accounts.

File: applications/nftaa.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (PieWol) [reply applications/nftaa.md:61]: thanks, I'd love that. Utility sounds reasonable.

File: applications/nftaa.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (PieWol) [reply applications/nftaa.md:78]: I don't see anything about chain-extensions mentioned in the doc and I think this is a deprecated feature. Would be great if you can check on this. Also I think if you were to realize this on the smart-contract level then you can only have this logic be applied on the smart contract level. You wouldn't be able to influence the chain level. Please correct me if I'm wrong.

File: applications/nftaa.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (PieWol) [reply applications/nftaa.md:87]: We already have the identity pallet which is able to represent additional information linked to an account. Given that this is the established way of linking information to an account I suggest to re-evaluate this strategy. This NFT attribute approach would introduce unnecessary overhead to existing projects that aggregate data linked to accounts. e.g. Polkassembly.

File: applications/nftaa.md (hunk 9 above)
Comment (Roman-24) [reply applications/nftaa.md]: You're right. A change has already been made and the lightweight application is no longer part of the proposal.
You can see the changes [here](https://github.com/w3f/Grants-Program/pull/2396/commits/1a8e20e1d464cf4d7e29e3da464d0673f3ca191d)

File: applications/nftaa.md (hunk 8 above)
Comment (PieWol) [reply applications/nftaa.md]: nice.

**Review by PieWol — CHANGES_REQUESTED** [review]
Thanks for your comments. I answered or marked the conversations as resolved. Looking forward to your answers to all unresolved comments. I appreciate the time and effort you are putting into this to explain your project and that you are incorporating our feedback 🙏

File: applications/nftaa.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (Roman-24) [reply applications/nftaa.md:78]: It looks like you are correct, anyway what is stated in the proposition remains valid because to integrate this solution to other pallets, our preferred way is through the pallet approach. But definitely, we will need to freshen up our state of the art in our minds based on our chain extension not knowledge :))))

File: applications/nftaa.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (Roman-24) [reply applications/nftaa.md:87]: Thanks for this comment and I'm sure it can be used since NFTAA will be like an account so it shouldn't be a problem to join. However, in our case we were thinking more of the technical attributes but thus the detail will only be clear when implemented. We just wanted to have the "corridor" more open so we don't promise something which could be potentially improved.

File: applications/nftaa.md (hunk 5 above)
Comment (Roman-24) [reply applications/nftaa.md:188]: for example, one possibility is: If NFTAA is listed you will not be able to sell assets owned by NFTAA

**Comment by Roman-24** [comment]
Thank you for reviewing and resolving the conversations. We appreciate your feedback and have addressed all unresolved comments. Your insights were valuable in refining the project, and we’re grateful for your collaboration.

File: applications/nftaa.md (hunk 5 above)
Comment (PieWol) [reply applications/nftaa.md:188]: That sounds good 👍  I'd like to see that. Would you be willing to integrate this into the deliverables?

File: applications/nftaa.md (hunk 7 above)
Comment (PieWol) [reply applications/nftaa.md:366]: I see. Then just quickly could you please describe how this showcase will look like? Will it be a webapp that you create so that users can conveniently interact with their NFTAAs for staking? I initially understood this milestone in a way that you are planning to introduce additional staking related logic within your NFTAA pallet. I wouldn't like this for the aforementioned reasons in my earlier comments. If you are planning to showcase NFTAAs for staking with a web-app unrelated to the pallet, then I'd be fine with that.

**Comment by PieWol** [comment]
Hey @Roman-24, it's only two more comments from my side. I hope we will have all items resolved soon. I'll also ping the committee about your application again later today so that you will receive more reviews. Sorry for the delay.

File: applications/nftaa.md (hunk 7 above)
Comment (Roman-24) [reply applications/nftaa.md]: I rearranged it all in my head a little bit and I think we finally found the same point of view. It will be as you write we will use staking as a showcase using the web app but we won't add any specific functionality to the nftaa_pallet itself for this functionality. So our generic proxy functionality will be used for staking. For that reason I have removed this parts from the milestone:
- 2c. nftaa stake/unstake
- 2d. nftaa increase/decrease

Thank you for your patience in this subdiscussion :)

File: applications/nftaa.md (hunk 5 above)
Comment (Roman-24) [reply applications/nftaa.md:188]: Ofc, I added it to [milestone 1](https://github.com/w3f/Grants-Program/blob/4bdd90ff5dc30e1d4db13d0701df2568c1c773bc/applications/nftaa.md#milestone-1--basic-functionality) in point 1d

**Review by semuelle — APPROVED** [review]

File: applications/nftaa.md (hunk 7 above)
Comment (PieWol) [reply applications/nftaa.md]: got it, nice!

File: applications/nftaa.md (hunk 1 above)
Comment (PieWol) [reply applications/nftaa.md:140]: Sounds good, maybe you can find a way to simply distribute these NFTs from one chain to other chains to have the effect of NFTAAs on all relevant chains. Something to figure out later.

**Comment by PieWol** [comment]
Hey @Roman-24 ,
thanks for resolving all questions and comments. I have one final question about your plans for this project once the grant ends. What will happen? What is your motivation to create this in the first place? Will you try to keep maintaining it? Will you publish the pallets as crates to be re-used by other teams? I'd love to hear some more about your future plans. I think after all the effort you put into this application it's worthy a few more words than what is currently listed under ``future plans`` in the application. Thanks.

**Comment by Roman-24** [comment]
@PieWol 
As I also mentioned above when discussing with semuelle - our goal is to expand the boundaries of how web3 uses NFTs today. It's an extension of how current NFTs are used, adding a whole new set of capabilities. This is the first place motivation. 

As we mentioned in future plans we plan to continue the maintenance and of course if something makes sense to us again we will bring further extensions. We are also aware that we will need to communicate with users and other teams after the end of the grant to ensure a better life for this project :).

Yes, we’re enthusiastic about publishing pallets as crates to enable reuse by other teams.

Is there anything you would recommend for the future of the project?

**Comment by PieWol** [comment]
Okay, thanks for clarifying. You have my approval :)

**Review by PieWol — APPROVED** [review]

**Review by keeganquigley — APPROVED** [review]
Thanks for the changes @Roman-24 I'm willing to give it a shot.

**Review by Noc2 — APPROVED** [review]
